x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -304,6 +338,80 @@ static void getMemoryAndProcessors(
         );
     }
 
+    /**
+     * Check if the dummy autoscaling entity task can be added by placing
+     * the task on the least loaded node.
+     *
+     * If there exists a node that can accommodate the dummy entity then return true (nothing to do),
+     * else return false and increment the memory and processor counts accordingly.
+     *
+     * We perform the calculation by identifying the least loaded node in terms of memory
+     * and determining if the addition of the dummy entity's memory and processor requirements could
+     * be accommodated on it.
+     *
+     * If the calculation returns false then treat the case as for a single trained model job
+     * that is already assigned, i.e. increment modelMemoryBytesSum and processorsSum appropriately.
+     *
+     * @param perNodeJobRequirements per Node lists of requirements
+     * @param perNodeMemoryInBytes total model memory available on every node
+     * @param perNodeProcessors total processors on every node
+     * @param dummyAutoscalingEntity "dummy" entity requirements used to potentially trigger a scaling event
+     * @return true if the dummy entity can be accommodated, false if not
+     */
+    static boolean dummyEntityFitsOnLeastLoadedNode(
+        Map<String, List<MlJobRequirements>> perNodeJobRequirements, // total up requirements...
+        long perNodeMemoryInBytes,
+        int perNodeProcessors,
+        MlDummyAutoscalingEntity dummyAutoscalingEntity
+    ) {
+
+        if (dummyAutoscalingEntity.processors == 0 && dummyAutoscalingEntity.memory == 0L) {
+            return true;
+        }
+
+        if (perNodeJobRequirements.size() < 1) {
+            return false;
+        }
+
+        Optional<MlJobRequirements> leastLodedNodeRequirements = Optional.ofNullable(
+            perNodeJobRequirements.entrySet()
+                .stream()
+                .map(
+                    entry -> tuple(
+                        entry.getKey(),
+                        entry.getValue()
+                            .stream()
+                            .reduce(
+                                MlJobRequirements.of(0L, 0, 0),
+                                (subtotal, element) -> MlJobRequirements.of(
+                                    subtotal.memory + element.memory,
+                                    subtotal.processors + element.processors,
+                                    subtotal.jobs + element.jobs
+                                )
+                            )
+                    )
+                )
+                .min(Comparator.comparingLong(tuple -> tuple.v2().memory))
+                .get()
+                .v2()
+        );
+
+        assert (leastLodedNodeRequirements.isPresent());
+        assert leastLodedNodeRequirements.get().memory >= 0L;
+        assert leastLodedNodeRequirements.get().processors >= 0;
```

Comment:
Are these definitely impossible? I think we could have an ML node with nothing running on it, and then these two variables would legitimately be 0.

Reply:
I _think_ these checks are legitimate as-is. They're checking against the presence of negative values, which should never be the case. The check for the memory requirements is the same as in `checkIfOneNodeCouldBeRemoved`.

Reply:
D'oh! I misread as `> 0`. 🤦